File: src/converter/plugins/convertArrayType.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";

export const convertArrayType = createSimplePlugin((node, context, render) => {
    if (
        ts.isTypeOperatorNode(node) &&
        node.operator === ts.SyntaxKind.ReadonlyKeyword &&
        ts.isArrayTypeNode(node.type)
    ) {
        const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
        checkCoverageService?.cover(node)
        checkCoverageService?.cover(node.type)

        return `js.array.ReadonlyArray<${render(node.type.elementType)}>`
    }

    if (ts.isArrayTypeNode(node)) {
        const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
        checkCoverageService?.cover(node)

        return `js.array.ReadonlyArray<${render(node.elementType)}>`
    }

    return null
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertArrayType.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.SyntaxKind
import typescript.isArrayTypeNode
import typescript.isTypeOperatorNode

val convertArrayType = createPlugin plugin@{ node, context, render ->
    if (
        isTypeOperatorNode(node) &&
        node.operator === SyntaxKind.ReadonlyKeyword
    ) {
        val type = node.type
        if (isArrayTypeNode(type)) {
            val checkCoverageService = context.lookupService(checkCoverageServiceKey)
            checkCoverageService?.cover(node)
            checkCoverageService?.cover(type)

            return@plugin "js.array.ReadonlyArray<${render(type.elementType)}>"
        }
    }

    if (isArrayTypeNode(node)) {
        val checkCoverageService = context.lookupService(checkCoverageServiceKey)
        checkCoverageService?.cover(node)

        return@plugin "js.array.ReadonlyArray<${render(node.elementType)}>"
    }

    null
}
